### PR TITLE
[release/8.0] [Blazor] Fix for URLs/querystrings not updating for interactive components when using SSR routing

### LINKED ISSUE #51216: NavigationManager.Uri does not update for interactive components on non-interactive routers
Repro:

* Create a Blazor Web project with the default options (interactivity=server, type=per-page)
* In `Counter.razor`, add `@inject NavigationManager Nav` and in the UI display `@Nav.Uri`
* In `Home.razor`, add `@attribute [RenderModeInteractiveServer]`
* Launch the app so you're first on the home screen, then navigate to counter

Expected: Displays the URL `https://.../counter`

Actual: Displays the URL `https://.../`

If you do an F5 refresh there, it will display the correct URL. But if you then navigate to other interactive pages, they still think the URL is `/counter`.

### Also affects [SupplyParameterFromQuery]

Similarly if you add code like this to `Counter.razor`:

```razor
[SupplyParameterFromQuery] public int? IncrementAmount { get; set; }

private void IncrementCount()
{
    currentCount += IncrementAmount.GetValueOrDefault(1);
}
```

... and change `NavMenu.razor` to contain `<NavLink class="nav-link" href="counter?incrementamount=10">`, then you'll find that:

 * If you start on the home screen and navigate to counter, it only increments in 1s when it should be 10s
 * If you do an F5 refresh while at `/counter?incrementamount=10`, it will then increment in 10s

This also means that if `Counter.razor` contained links to itself but with different `incrementamount` query values (e.g., to let the user change to their preferred amount), the links simply won't have any effect.

I discovered this while trying to implement paging for a grid that should support both static SSR and interactive modes.

### PR DESCRIPTION
**Pull request by MackinnonBuck**
# Invoke `LocationChanged` for interactive components when an enhanced navigation occurs

This PR makes the `LocationChanged` event get invoked for interactive runtimes when an enhanced navigation occurs.

## Description

This PR makes the following changes:
* The `LocationChanged` event gets invoked for all interactive runtimes when either a client-side or enhanced navigation occurs
  * Previously:
    * An enhanced navigation would have not invoked the `LocationChanged` event in interactive runtimes
    * A client-side navigation would only invoke the `LocationChanged` event for the runtime that started most recently
* Attempting to enable interactive routing from multiple interactive runtimes results in an error at runtime
* If an interactive router is present but multiple runtimes are available, the `LocationChanging` event only gets invoked for the runtime that has the interactive router
  * It would take additional work to allow either runtime to cancel a client-side navigation, and that's outside the scope of this PR. We have https://github.com/dotnet/aspnetcore/issues/48766 to track doing the extra work to support this scenario

Fixes #51216

## Customer Impact

Without this changes the customers will encounter problems with UI being correctly updated during enhanced navigation because `LocationChanging` event not firing. 

Besides the UI not updating correctly this will also cause customer registered handlers for the `LocationChanged` event not firing resulting in unexpected behaviors.

## Regression?

- [ ] Yes
- [x] No

[If yes, specify the version the behavior has regressed from]

## Risk

- [ ] High
- [ ] Medium
- [x] Low

Despite the size of this change, we have quite extensive automated tests (including these newly added ones) around the interactive routing scenarios, so we expect that nothing will break from this change.

## Verification

- [x] Manual (required)
- [x] Automated

## Packaging changes reviewed?

- [ ] Yes
- [x] No
- [ ] N/A

----

## When servicing release/2.1

- [ ] Make necessary changes in eng/PatchConfig.props

### CODE REVIEW

File: src/Components/Web.JS/src/Services/NavigationManager.ts [hunk 1]
```diff
@@ -4,16 +4,22 @@
...
-let notifyLocationChangedCallback: ((uri: string, state: string | undefined, intercepted: boolean) => Promise<void>) | null = null;
-let notifyLocationChangingCallback: ((callId: number, uri: string, state: string | undefined, intercepted: boolean) => Promise<void>) | null = null;
+type NavigationCallbacks = {
+  rendererId: WebRendererId;
+  hasLocationChangingEventListeners: boolean;
+  locationChanged(uri: string, state: string | undefined, intercepted: boolean): Promise<void>;
+  locationChanging(callId: number, uri: string, state: string | undefined, intercepted: boolean): Promise<void>;
+};
+
+const navigationCallbacks = new Map<WebRendererId, NavigationCallbacks>();
```
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Services/NavigationManager.ts:22]: This state was previously shared between all interactive runtimes. Whichever runtime called `listenForNavigationEvents` would overwrite the state of the previous runtime.

This change allows multiple runtimes to have their own callbacks registered simultaneously.

File: src/Components/Web.JS/src/Services/NavigationManager.ts [hunk 2]
```diff
@@ -216,18 +236,12 @@ function ignorePendingNavigation() {
...
 
-function notifyLocationChanging(uri: string, state: string | undefined, intercepted: boolean): Promise<boolean> {
+function notifyLocationChanging(uri: string, state: string | undefined, intercepted: boolean, callbacks: NavigationCallbacks): Promise<boolean> {
   return new Promise<boolean>(resolve => {
     ignorePendingNavigation();
-
-    if (!notifyLocationChangingCallback) {
-      resolve(false);
-      return;
-    }
```
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Services/NavigationManager.ts:226]: I'm pretty sure this condition was just here so the TypeScript compiler wouldn't complain about `notifyLocationChangingCallback` being undefined. So I don't think we're losing any meaningful behavior by removing this.

**Comment by SteveSandersonMS** [comment]
This is looking great.

@surayya-MS Am I right to think that the only missing piece remaining is adding a Selenium test that makes use of the `InteractiveNavigationComponent` to show that `Navigation.Uri` gets updated when there's an enhanced nav to a different host page?

It would be good also to show that `[SupplyParameterFromQuery]` also updates inside the interactive component.

**Review by SteveSandersonMS — APPROVED** [review]
Fantastic! Thanks for handling all this, @surayya-MS and @MackinnonBuck!

**Pull request merged**